plugins/ssrmatching/ssrmatching.ml

```diff
@@ -606,7 +607,11 @@ let match_upats_HO ~on_instance upats env sigma0 ise c =
             (EConstr.push_rel (Context.Rel.Declaration.LocalAssum(x, t)) env)
             ise' pb b
         | KpatFlex | KpatProj _ ->
-          unif_HO env ise u.up_f (mkSubApp f (i - Array.length u.up_a) a)
+          let tuf = Retyping.get_type_of ~lax:true env ise u.up_f in
+          let fa = mkSubApp f (i - Array.length u.up_a) a in
+          let tfa = Retyping.get_type_of ~lax:true env ise fa in
+          let ise = unif_HO env ise tuf tfa in
```

Comment:
@proux01 this part is really necessary only if a primproj is involved, IIRC

Reply:
Yes but isn't it always the case under KpatProj?

Reply:
`KpatProj` should mean, IIRC, a canonical record projection. Primitive or not.

I think the retyping is only necessary in the primitive case, and maybe not always. This is why the code we tried the other day way trying to unify twice, one upfront and one after retyping in case of failure (and if the terms were primprojs).

Then, if there is no slowdown, I'm fine with this code, but I doubt so.

Reply:
I did not notice any slowdown on a full compilation of math-comp but I didn't bench very carefully to be honest.

Reply:
Let's be careful anyway.